src/debug_training/ROS_Tools.md

```diff
@@ -0,0 +1,316 @@
+# WRoverPlayground - ROS Tools
+
+## ROS Concepts Crash Course
+
+ROS nodes are processes that use ROS to communicate with other nodes.
+ROS nodes communicate with each other using streaming topics, RPC services, and the Parameter Server.
+
+ROS topics are named interfaces with a publisher and subscriber model.
+Publisher nodes generate data and publish that data to a topic.
+Subscriber nodes subscribe to a topic to use data that is published to a topic.
+A topic can have multiple publisher and subscriber nodes.
+Topics are intended for unidirectional communication.
+If a response is expected from a request, a service should be used instead.
+
+ROS services allow nodes to return data to other nodes.
+Server nodes define the service request and response.
+Client nodes call the service by sending a request and await the response.
+
+Both topics and services are defined using messages.
+A message is a data structure containing typed fields.
+For a service, a request and a response message must be defined.
+
+ROS topics and services will be covered in more detail in the next training modules.
+
+The final method of communication between ROS nodes is the Parameter Service.
+This is essentially a shared dictionary that can be used to store and retrieve variables at runtime.
+It is usually used for configuration since it is not designed for high performance.
+
+## ROS Command Line
+
+### Starting nodes
+
+Open a terminal window and run `roscore`.
+This starts the ROS master node.
+
+Open a new terminal and run `rosnode list`.
+You should see a single node: `/rosout`.
+
+Run `rosnode info /rosout`.
+You should see the following output: 
+```
+Node [/rosout]
+Publications: 
+ * /rosout_agg [rosgraph_msgs/Log]
+
+Subscriptions: 
+ * /rosout [unknown type]
+
+Services: 
+ * /rosout/get_loggers
+ * /rosout/set_logger_level
+```
+This shows that the `/rosout` node publishes to the `/rosout_agg` topic and subscribes sto the `/rosout` topic.
+Also, the `/rosout` node provides the `/rosout/get_loggers` and `/rosout/set_logger_level` services.
+
+Now, let's start another node.
+Run `rosrun turtlesim turtlesim_node`.
```

Comment:
Does turtlesim come from installing ROS, or do new members need to do another installation?

Reply:
Turtlesim is included in the `osrf/ros:noetic-desktop-full` docker image. I'll add a note that it might need to be installed with `sudo apt install ros-noetic-ros-tutorials` if not using docker.